examples/pytorch/rgcn/link_predict_mp.py

```diff
@@ -0,0 +1,1147 @@
+                                    pin_memory=True,
+                                    drop_last=False,
+                                    num_workers=0)
+            for seeds, blocks in tqdm.tqdm(dataloader):
+                block = blocks[0]
+                block = block.int().to(device)
+                idx = block.srcdata[dgl.NID]
+                h = layer(block, x[idx].to(device), block.edata['etype'], block.edata['norm'])
+                y[seeds] = h.cpu()
+
+            x = y
+        return y
+
+def regularization_loss(n_emb, r_emb):
+    return th.mean(n_emb.pow(2)) + \
+            th.mean(r_emb.pow(2))
+
+def calc_pos_score(h_emb, t_emb, r_emb):
+    # DistMult
+    score = th.sum(h_emb * r_emb * t_emb, dim=-1)
+    return score
+
+def calc_neg_tail_score(heads, tails, r_emb, num_chunks, chunk_size, neg_sample_size, device=None):
+    hidden_dim = heads.shape[1]
+    r = r_emb
+
+    if device is not None:
+        r = r.to(device)
+        heads = heads.to(device)
+        tails = tails.to(device)
+    tails = tails.reshape(num_chunks, neg_sample_size, hidden_dim)
+    tails = th.transpose(tails, 1, 2)
+    tmp = (heads * r).reshape(num_chunks, chunk_size, hidden_dim)
+    return th.bmm(tmp, tails)
+
+def calc_neg_head_score(heads, tails, r_emb, num_chunks, chunk_size, neg_sample_size, device=None):
+    hidden_dim = tails.shape[1]
+    r = r_emb
+    if device is not None:
+        r = r.to(device)
+        heads = heads.to(device)
+        tails = tails.to(device)
+    heads = heads.reshape(num_chunks, neg_sample_size, hidden_dim)
+    heads = th.transpose(heads, 1, 2)
+    tmp = (tails * r).reshape(num_chunks, chunk_size, hidden_dim)
+    return th.bmm(tmp, heads)
+
+def get_loss(h_emb, t_emb, nh_emb, nt_emb, r_emb,
+    num_chunks, chunk_size, neg_sample_size):
+    # triplets is a list of data samples (positive and negative)
+    # each row in the triplets is a 3-tuple of (source, relation, destination)
+    pos_score = calc_pos_score(h_emb, t_emb, r_emb).view(-1, 1)
+    t_neg_score = calc_neg_tail_score(h_emb, nt_emb, r_emb, num_chunks, chunk_size, neg_sample_size)
+    h_neg_score = calc_neg_head_score(nh_emb, t_emb, r_emb, num_chunks, chunk_size, neg_sample_size)
+    #pos_score = F.logsigmoid(pos_score)
+    h_neg_score = h_neg_score.view(-1, 1)
+    t_neg_score = t_neg_score.view(-1, 1)
+    #h_neg_score = F.logsigmoid(-h_neg_score).mean(dim=1)
+    #t_neg_score = F.logsigmoid(-t_neg_score).mean(dim=1)
+
+    score = th.cat([pos_score, h_neg_score, t_neg_score])
+    label = th.cat([th.full((pos_score.shape), 1.0),
+                    th.full((h_neg_score.shape), 0.0),
+                    th.full((t_neg_score.shape), 0.0)]).to(score.device)
+    predict_loss = F.binary_cross_entropy_with_logits(score, label)
+    #pos_score = pos_score.mean()
+    #h_neg_score = h_neg_score.mean()
+    #t_neg_score = t_neg_score.mean()
+    #predict_loss = -(2 * pos_score + h_neg_score + t_neg_score) / 2
```

Comment:
Are these comment-out lines useful?

Reply:
Yes, different ways of achieving the loss.
Maybe we can remove it.